### ISSUE TITLE
capng_change_id() can leave PR_SET_KEEPCAPS set to 1 on failure

### ISSUE THREAD

**Issue by BtbN** [stevegrubb/libcap-ng#33]
While working on adding capability support for openvpn, we discovered a very problematic possible failure case of capng_change_id():

In case capng_update_id() fails, maybe due to missing capabilities, rampant SElinux, or whatever other reason, we fall back to the classic codepath, which just calls setgid() + setuid().
Now, if capng_change_id() fails after https://github.com/stevegrubb/libcap-ng/blob/423a1afe5e6c11442b923e641368e84e0d271bab/src/cap-ng.c#L994 but before actually having changed the user in https://github.com/stevegrubb/libcap-ng/blob/423a1afe5e6c11442b923e641368e84e0d271bab/src/cap-ng.c#L1041 we are left with PR_SET_KEEPCAPS set.

This results in our fallback setuid() leaving the permitted capabilities set. This is a very unexpected and potentially dangerous if it's a security critical process that is expected to have no capabilities.

Is this behaviour expected? As in, we should always terminate in case capng_change_id() failed, since the state cannot be guaranteed?
Or would it be possible to reset the PR_SET_KEEPCAPS flags on failure? Or maybe set it much later, so there's not so much stuff that can fail in between it and the actual setuid?

**Comment by stevegrubb** [comment]
The man page says that probably the best thing to do on any failure of this function is to exit. You'd need to re-read the process capabilities to assess what the security posture is since this function could have overwritten everything.

That said, there are people that don't check the return codes. So, it might not be a bad idea to reset PR_SET_KEEPCAPS on the way out.

**Comment by stevegrubb** [comment]
Commit a458292 should address this.

**Comment by BtbN** [comment]
Yeah, we figured that the only perfect way to handle a failure was to exit.
But the extra bit of hardening is definitely a good idea either way.

In the case of OpenVPN all it tries in case of failure is to drop all permissions instead of retaining some, if it still can, as backwards-compat to before capability awareness was added. So either that still works, or capabilities are too messed up and it fails as well, which will lead to an a fatal error and exit.

Commit looks good on first glance to me, thanks!
